### PR TITLE
Don't allow replacement of root develop specs with --reuse

### PR DESCRIPTION
**Pull request by psakievich**
Based off conversations with @tgamblin, this makes it so `--reuse` will never replace a develop spec since we always want to build those.

### CODE REVIEW

**Comment by haampie** [comment]
Does not work? Or do I misunderstand "you can't choose an installed hash for a dev spec"?

```
$ cat spack.yaml
spack:
  specs: [zlib]
  view: false
  develop:
    zlib:
      spec: zlib@1.2.11
  config:
    install_tree:
      root: ./store
$ spack -e . install
$ spack -c config:install_tree:root:$PWD/store spec -I --reuse zstd+programs
Input spec
--------------------------------
 -   zstd+programs

Concretized
--------------------------------
 -   zstd@1.5.0%gcc@10.3.0+programs arch=linux-ubuntu20.04-zen2
 -       ^lz4@1.9.3%gcc@10.3.0 libs=shared,static arch=linux-ubuntu20.04-zen2
 -       ^lzma@4.32.7%gcc@10.3.0 arch=linux-ubuntu20.04-zen2
[+]      ^zlib@1.2.11%gcc@10.3.0+optimize+pic+shared dev_path=/tmp/tmp.edlFqyZ9yr/zlib arch=linux-ubuntu20.04-zen2
```

**Comment by tgamblin** [comment]
It’s ok to reuse *dependencies* of develop specs if they satisfy the requirements. What @psakievich was observing was that the concretizer would reuse the root, which currently has the issue that reused specs don’t have their build dependencies attached (we currently strip them and don’t include them in the db, for reasons mentioned in #28504, which will fix that).

What this PR is saying is that dev roots should always be built, and therefore should have their build dependencies concretized.

After #28504, we may want to rethink the semantics a bit and allow some reuse for dev specs, but I can also see a case for this still being needed. I think people may want the roots to always be concretizer “fresh” — this gets into some of the points @adamjstewart raised in #28468.

**Comment by haampie** [comment]
What's special about root specs? I know people developing packages {x, y, z} at the same time, where x depends on y depends on z.

**Comment by tgamblin** [comment]
I should be more clear. I mean the roots of dev specs — so whatever node (zstd here) you said was a develop spec (as opposed to the whole sub-DAG). Technically that doesn’t have to be an environment root. I mean the root of the develop spec’s sub-DAG. A develop spec can be an env root or some dependency of one as you say.

What is special about dev specs is that they need to be built, and right now if you reuse, we pull the spec in by hash from the DB, and you lose build dependencies like CMake, which you need for, well, building.

**Comment by haampie** [comment]
Right, but suppose you do `spack install --reuse x`, and `x` depends on `y`, and there's a development version of `y` in the database. Do you mean that Spack should now include `y`'s build deps and rebuild (i.e. effectively do install --overwrite) if changed according to heuristics? In particular outside the env that has the `develop:` section?

To me it makes sense that dev specs get rebuilt inside the environment that marks those specs as `develop`. But outside of an environment: either you reuse the dev spec as is (so no build deps are required), or you don't consider the dev spec at all because during concretization it's unclear whether or not it can be reused without building from sources?

**Comment by tgamblin** [comment]
> Do you mean that Spack should now include y's build deps and rebuild (i.e. effectively do install --overwrite) if changed according to heuristics?

Good question. All this is saying is that y will be concretized *without considering* installations. If it ends up with the same DAG hash as the prior installation of y, it won’t rebuild, it’ll reuse by exact hash match like always, unless you ask for a rebuild with —overwrite. But the spec output from the concretizer will have build dependencies, so if we *need* a rebuild, it’ll work.

This is really just a workaround for the lack of build deps in the DB, and the fact that —reuse reads specs back from the DB (without build deps). #28504 will make this much easier to think about because we won’t strip build deps anymore.

**Comment by psakievich** [comment]
> To me it makes sense that dev specs get rebuilt inside the environment that marks those specs as `develop`. But outside of an environment: either you reuse the dev spec as is (so no build deps are required), or you don't consider the dev spec at all because during concretization it's unclear whether or not it can be reused without building from sources?

@haampie this is a good question. I've been thinking about this as well.  My current strategy/thinking would lean toward never re-using dev_builds in other environments.  If I really want to reuse it I will specify its install path in an external, or providing a specific way through spack to intentionally reuse them.  I think allowing the concretizer to just grab a dev spec is rather dangerous.  How do I know if it is grabbing an install of a feature branch I was developing unintentionally?  I'm just starting to learn about the concretizer so maybe it is taken care of, but this seems like a danger to me in the `--reuse` paradigm. 

TLDR: I think the reuse of dev_specs should always be an explicit user decision and not something the concretizer can decide on its own.

**Comment by haampie** [comment]
The TL;DR is what I assumed this PR was supposed to fix, hence the example in my first comment :laughing:

**Comment by psakievich** [comment]
@haampie yeah that is a problem and it would be good to fix it here. I'm but an infant in the world clingo, but I can see if I can figure it out. This is currently breaking my workflow so I don't want to leave it open for too long without a merge.  If not we should post an issue to make sure it gets addressed.  I posted this PR just to get it up and on radars.  I still need to add tests for it so it's not ready to merge yet anyways.

**Review by becker33 — DISMISSED** [review]
I think we can keep this separate from the other concerns raised here

**Comment by psakievich** [comment]
I updated the title to be more specific. Not sure why the tests got canceled on this one.  Also I need a little guidance on where to look for writing a test for this case.  @becker33 or @tgamblin could you point me to a file that tests the concretizer. Would this fall into `concretize_preferences.py`?

**Comment by psakievich** [comment]
@becker33 and @tgamblin what are your thoughts on merging this? I know @tgamblin wanted a test.  I'm not really familiar with testing the concretizer though. Given the simplicity of this change would you consider merging it without a new test? I'll write one if we need it, but I need a little guidance on where to look for concretizer tests.

**Comment by becker33** [comment]
@haampie I think the newest commit should solve the problem you were seeing

File: lib/spack/spack/test/concretize.py [hunk 1]
```diff
@@ -1138,14 +1137,46 @@ def test_external_package_versions(self, spec_str, is_external, expected):
...
+        package_version = '0.0.0'
+        spec_str = package_name
+        dev_spec_str = '%s@%s dev_path=/fake' % (package_name,
+                                                 package_version)
+        # concretize and install a non-dev version
+        s = Spec(spec_str).concretized()
+        s.package.do_install(fake=True)
+        # concretize a dev version
+        dev_spec = Spec(dev_spec_str).concretized(reuse=True)
+        assert dev_spec.dag_hash() is not s.dag_hash()
```
Comment (becker33) [inline lib/spack/spack/test/concretize.py]:
```suggestion
        assert dev_spec.dag_hash() != s.dag_hash()
```
`is not` checks whether they are the same string object in memory, `!=` checks equality.

File: lib/spack/spack/test/concretize.py [hunk 2]
```diff
@@ -1138,14 +1137,46 @@ def test_external_package_versions(self, spec_str, is_external, expected):
...
+
+        spec_str = root_spec
+        dev_spec_str = '%s ^%s@%s dev_path=/fake' % (root_spec, package_name,
+                                                     package_version)
+        # concretize and install a non-dev version
+        s = Spec(spec_str).concretized()
+        s.package.do_install(fake=True)
+        # concretize a dev version
+        dev_spec = Spec(dev_spec_str).concretized(reuse=True)
+        assert dev_spec.dag_hash() is not s.dag_hash()
```
Comment (becker33) [inline lib/spack/spack/test/concretize.py]:
```suggestion
        assert dev_spec.dag_hash() != s.dag_hash()
```
Same here

**Review by becker33 — CHANGES_REQUESTED** [review]
I think these two comparison changes should fix the issue where these tests passed without the fix.

Edit to add:

I think the test will be more robust if you remove everything that pins the version. Because you're pinning the dev version but not the normal spec, if the default version of the package changes the test will no longer be testing what it needs to test.

Dev specs typically have to include a version from the user interface, but there's nothing on the spec object enforcing that, so you don't need it to make the test work.

File: lib/spack/spack/test/concretize.py (hunk 1 above)
Comment (psakievich) [reply lib/spack/spack/test/concretize.py]: 🤔 thanks!

**Comment by psakievich** [comment]
Thanks @becker33 yes that is the only reason I added the versions.  I will remove them and make the other changes

**Comment by psakievich** [comment]
I don't understand why codecov is failing. I only added tests. 🤦‍♂️

**Review by becker33 — APPROVED** [review]
